File: app/src/main/java/com/app/memoeslink/adivinador/ApplicationContext.java
package com.app.memoeslink.adivinador;

import android.preference.PreferenceManager;

import androidx.multidex.MultiDexApplication;

import net.danlew.android.joda.JodaTimeAndroid;

import org.apache.commons.lang3.StringUtils;

import java.io.File;

public class ApplicationContext extends MultiDexApplication {
    private SharedPreferencesHelper preferences;
    private SharedPreferencesHelper defaultPreferences;

    @Override
    public void onCreate() {
        super.onCreate();
        preferences = new SharedPreferencesHelper(this, SharedPreferencesHelper.PREFERENCES);
        defaultPreferences = new SharedPreferencesHelper(this);

        //Initialize libraries
        JodaTimeAndroid.init(this);

        //Set preference default values
        PreferenceManager.setDefaultValues(ApplicationContext.this, R.xml.default_preferences, false);

        //Delete old databases
        deleteOldDatabases();
    }

    public File getDatabaseTrace(int version) {
        if (version >= preferences.getInt("revisedDatabaseVersion", 1)) {
            String databaseName;

            if (version > 1)
                databaseName = String.format(DatabaseConnection.DATABASE_NAME_FORMAT, "_upgrade_" + (version - 1) + "-" + version);
            else
                databaseName = String.format(DatabaseConnection.DATABASE_NAME_FORMAT, "");
            File database = getDatabasePath(databaseName);

            if (database.exists()) {
                preferences.putInt("revisedDatabaseVersion", version);
                return database;
            }
            getDatabaseTrace(version - 1);
        }
        return null;
    }

    private File[] getDatabaseList() {
        try {
            File database = getDatabaseTrace(DatabaseConnection.DATABASE_VERSION);

            if (database != null) {
                File directory = database.getParentFile();
                System.out.println("Database directory: " + directory);
                File[] files = directory.listFiles();
                System.out.println("Number of files in database directory: " + files.length);

                for (int i = 0; i < files.length; i++) {
                    System.out.println("Database (" + (i + 1) + "): " + files[i].getName());
                }
                return files;
            }
            System.out.println("There's not any database.");
        } catch (Exception ignored) {
        }
        return null;
    }

    public void deleteOldDatabases() {
        File[] files = getDatabaseList();

        if (files != null && files.length > 0) {
            for (File file : files) {
                if (!StringUtils.startsWith(file.getName(), "google_") && !file.getName().equals(DatabaseConnection.DATABASE_NAME)) {
                    if (file.exists()) {
                        if (file.delete())
                            System.out.println("Database was successfully deleted: " + file.toURI());
                        else
                            System.out.println("Database couldn't be deleted: " + file.toURI());
                    }
                }
            }
        } else
            System.out.println("There wasn't any old database to delete.");
    }
}

File: app/src/main/java/com/app/memoeslink/adivinador/ApplicationContext.kt
package com.app.memoeslink.adivinador

import androidx.multidex.MultiDexApplication
import androidx.preference.PreferenceManager
import com.memoeslink.generator.common.Session
import com.memoeslink.generator.common.StringHelper
import com.memoeslink.helper.SharedPreferencesHelper
import java.io.File

class ApplicationContext : MultiDexApplication() {
    private var preferences: SharedPreferencesHelper? = null
    private var defaultPreferences: SharedPreferencesHelper? = null

    override fun onCreate() {
        super.onCreate()
        preferences = SharedPreferencesHelper.getPreferencesHelper(this@ApplicationContext)
        defaultPreferences = SharedPreferencesHelper(this@ApplicationContext)

        //Initialize Generator database
        Session.getInstance().initDatabase(this@ApplicationContext)

        //Set preference default values
        PreferenceManager.setDefaultValues(
            this@ApplicationContext,
            R.xml.default_preferences,
            false
        )

        //Delete old databases
        deleteOldDatabases()
    }

    private fun getDatabaseTrace(version: Int): File? {
        if (version >= preferences?.getInt("revisedDatabaseVersion", 1) ?: 1) {
            val databaseName = if (version > 1) {
                String.format(
                    Database.DATABASE_NAME_FORMAT,
                    "_upgrade_" + (version - 1) + "-" + version
                )
            } else {
                String.format(Database.DATABASE_NAME_FORMAT, "")
            }
            val database = getDatabasePath(databaseName)

            if (database.exists()) {
                preferences?.putInt("revisedDatabaseVersion", version)
                return database
            }
            getDatabaseTrace(version - 1)
        }
        return null
    }

    private fun getDatabaseList(): Array<File>? {
        try {
            getDatabaseTrace(Database.DATABASE_VERSION)?.parentFile.let { directory ->
                val files = directory?.listFiles()
                println("Database directory: $directory")
                println("Number of files in database directory: " + files?.size)

                files?.forEachIndexed { index, file ->
                    println("Database (" + (index + 1) + "): " + file.name)
                }
                return files
            }
        } catch (ignored: Exception) {
            println("There's not any database.")
            return null
        }
    }

    private fun deleteOldDatabases() {
        val files = getDatabaseList()

        files?.takeUnless {
            files.isEmpty()
        }?.forEach { file ->
            if (!StringHelper.startsWith(
                    file.name,
                    "google_"
                ) && file.name != Database.DATABASE_NAME
            ) {
                when {
                    !file.exists() -> println("Database couldn't be found to be deleted: " + file.toURI())
                    !file.delete() -> println("Database couldn't be deleted: " + file.toURI())
                    else -> println("Database was successfully deleted: " + file.toURI())
                }
            }
        } ?: println("There wasn't any old database to delete.")
    }
}